modules/ruxhal/src/platform/aarch64_common/ivc.rs

```diff
@@ -0,0 +1,218 @@
+             control_table_ptr: ptr::null_mut(),
+         }
+     }
+ 
+     pub fn connect(&mut self, communication_zone: usize) -> Result<(), &'static str> {
+         let alloc_size = HVISOR_HC_IVC_INFO_SIZE;
+         let align = HVISOR_HC_IVC_INFO_ALIGN;
+         let layout = Layout::from_size_align(alloc_size, align).unwrap();
+         
+         let ptr = self.allocator.alloc(layout).expect("Memory allocation failed!");
+         self.ivc_info_ptr = ptr.as_ptr() as *mut IvCInfo;
+ 
+         let vpa_ivcinfo = self.ivc_info_ptr as usize;
+         let pa_ivcinfo: usize = vpa_ivcinfo - __PA;
+ 
+         debug!("The memory address of the IVC Info: VA: 0x{:x}, IPA: 0x{:x}", vpa_ivcinfo, pa_ivcinfo);
+ 
+         ivc_hvc_call(HVISOR_HC_IVC_INFO as u32, pa_ivcinfo, HVISOR_HC_IVC_INFO_SIZE);
+         debug!("ivc_hvc_call finished.");
+ 
+         // Safety: At this point we know ivc_info_ptr is valid and allocated
+         let ivc_info: &IvCInfo = unsafe { &*self.ivc_info_ptr };
+ 
+         let control_table_ptr = (ivc_info.ivc_ct_ipas[communication_zone] + __PA as u64) as *mut ControlTable;
+         self.control_table_ptr = control_table_ptr;
+ 
+         info!("IVC Connection established.");
+         Ok(())
+     }
+ 
+     pub fn send_message(&self, message: &str) -> Result<(), &'static str> {
+         if self.ivc_info_ptr.is_null() {
+             return Err("Not connected");
+         }
+     
+         let ivc_info: &IvCInfo = unsafe { &*self.ivc_info_ptr };
+     
+         let control_table: &ControlTable = unsafe { &*self.control_table_ptr };
+     
+         // Suppose we are zone1, we are to send message to zone0.
+         // Therefore use the out_sec_size field of the ControlTable struct (typically 0x1000).
+         let offset = control_table.out_sec_size as u64;
+         let address: u64 = ivc_info.ivc_shmem_ipas[0] + offset + __PA as u64;
+     
+         write_to_address(address, message)?;
+         info!("Message written to shared memory: {}", message);
+     
+         // Modify the control table to inform Zone0 Linux
+         let control_table: &mut ControlTable = unsafe { &mut *self.control_table_ptr };
+         debug!("Ipi_invoke reset to inform Zone0 linux.");
+         control_table.ipi_invoke = 0x0;
+     
+         Ok(())
+     }
+     
+ 
+     pub fn close(&mut self) -> Result<(), &'static str> {
+         if self.ivc_info_ptr.is_null() {
+             return Err("Not connected");
+         }
+         // free the memory
+         let layout = Layout::from_size_align(HVISOR_HC_IVC_INFO_SIZE, HVISOR_HC_IVC_INFO_ALIGN).unwrap();
+         self.allocator.dealloc(unsafe { NonNull::new_unchecked(self.ivc_info_ptr as *mut u8) }, layout);
+         info!("IVC Connection closed.");
+         Ok(())
+     }
+ }
+ 
+ /// ivc "hvc" method call
+ fn ivc_hvc_call(func: u32, arg0: usize, arg1: usize) -> usize {
+     let ret;
+     unsafe {
+         core::arch::asm!(
+             "hvc #4856",
```

Comment:
The "4856" in hvc #4856 is just a label value. The immediate value here does not affect the final execution result. The reason for using 4856 is that similar positions in the "hvisor-tool" project use 4856, and after asking, it was confirmed that there is no special meaning. It can be changed to 0, so let's change it to 0. In practical tests, the execution result with #0 is exactly the same.